pintc/src/predicate/optimize/const_folding.rs

```diff
@@ -33,13 +33,25 @@ pub(crate) fn const_folding(handler: &Handler, contract: &mut Contract) {
 /// Ex:
 /// 1 + 2 is 3
 /// 1 + 2 < 4 is true
+/// 1 + ::x is 6 where ::x is 5
 pub(crate) fn fold_consts(contract: &mut Contract) -> bool {
+    let mut scope_values: FxHashMap<String, Immediate> = FxHashMap::default();
     let mut expr_keys_to_replace: FxHashMap<ExprKey, (Expr, Type)> = FxHashMap::default();
     let mut did_fold_const: bool = false;
 
-    let evaluator = Evaluator::new(contract);
-
     for pred_key in contract.preds.keys().collect::<Vec<_>>() {
+        // Create a map of variables that were declared as immediates for folding later
+        if let Some(pred) = contract.preds.get(pred_key) {
+            for var in pred.variables() {
+                let expr = var.1.expr.get(contract);
+                if let Expr::Immediate { value, .. } = expr {
+                    scope_values.insert(var.1.name.clone(), value.clone());
+                }
+            }
```

Comment:
Something like this I find a bit more readable. I try to avoid `.0`, `.1`, etc. if possible because it's not immediately clear what they refer to. 

Also, I tend to avoid temporary variables if they're only used once and if they're not initialized with a "large" RHS.
```suggestion
            for (_, var) in pred.variables() {
                if let Expr::Immediate { value, .. } = var.expr.get(contract) {
                    scope_values.insert(var.1.name.clone(), value.clone());
                }
            }
```